src/quicklist.c

```diff
@@ -740,35 +738,40 @@ void quicklistDelEntry(quicklistIter *iter, quicklistEntry *entry) {
 void quicklistReplaceEntry(quicklistIter *iter, quicklistEntry *entry,
                            void *data, size_t sz)
 {
-    quicklist* quicklist = iter->quicklist;
+    quicklist *quicklist = iter->quicklist;
+    quicklistNode *node = entry->node;
 
-    if (likely(!QL_NODE_IS_PLAIN(entry->node) && !isLargeElement(sz))) {
-        entry->node->entry = lpReplace(entry->node->entry, &entry->zi, data, sz);
-        quicklistNodeUpdateSz(entry->node);
+    if (likely(!QL_NODE_IS_PLAIN(node) && !isLargeElement(sz) && 
+               (node->count == 1 || !quicklistNodeExceedsLimit(quicklist->fill, 
+                node->sz - entry->sz + sz + SIZE_ESTIMATE_OVERHEAD, node->count)))) {
+        node->entry = lpReplace(node->entry, &entry->zi, data, sz);
+        quicklistNodeUpdateSz(node);
         /* quicklistNext() and quicklistGetIteratorEntryAtIdx() provide an uncompressed node */
-        quicklistCompress(quicklist, entry->node);
-    } else if (QL_NODE_IS_PLAIN(entry->node)) {
+        quicklistCompress(quicklist, node);
+    } else if (node->count == 1) {
+        zfree(node->entry);
         if (isLargeElement(sz)) {
-            zfree(entry->node->entry);
-            entry->node->entry = zmalloc(sz);
-            entry->node->sz = sz;
-            memcpy(entry->node->entry, data, sz);
-            quicklistCompress(quicklist, entry->node);
+            node->entry = zmalloc(sz);
+            memcpy(node->entry, data, sz);
+            node->sz = sz;
+            node->container = QUICKLIST_NODE_CONTAINER_PLAIN;
         } else {
-            quicklistInsertAfter(iter, entry, data, sz);
-            __quicklistDelNode(quicklist, entry->node);
+            node->entry = lpPrepend(lpNew(0), data, sz);
+            quicklistNodeUpdateSz(node);
+            node->container = QUICKLIST_NODE_CONTAINER_PACKED;
         }
+        quicklistCompress(quicklist, node);
     } else {
-        entry->node->dont_compress = 1; /* Prevent compression in quicklistInsertAfter() */
-        quicklistInsertAfter(iter, entry, data, sz);
-        if (entry->node->count == 1) {
-            __quicklistDelNode(quicklist, entry->node);
+        /* node->count > 1, node will not be removed */
+        quicklistDelIndex(quicklist, node, &entry->zi);
+        iter->zi = NULL;
+        if (entry->offset == node->count) {
+            iter->offset--;
+            quicklistNext(iter, entry);
```

Comment:
I don't think we should engage in these dangerous behaviors.
`quicklistNext()` should be the responsibility of the caller of the iterator, and should not be used internally.
All we can do is update the iterator to the correct position when it changes or reset it to prevent it from being used again.

Reply:
It's removed, it causes bug anyway.